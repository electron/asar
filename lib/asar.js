'use strict'

const fs = require('./wrapped-fs')
const path = require('path')
const minimatch = require('minimatch')

const Filesystem = require('./filesystem')
const disk = require('./disk')
const crawlFilesystem = require('./crawlfs')

/**
 * Whether a directory should be excluded from packing due to the `--unpack-dir" option.
 *
 * @param {string} dirPath - directory path to check
 * @param {string} pattern - literal prefix [for backward compatibility] or glob pattern
 * @param {array} unpackDirs - Array of directory paths previously marked as unpacked
 */
function isUnpackedDir (dirPath, pattern, unpackDirs) {
  if (dirPath.startsWith(pattern) || minimatch(dirPath, pattern)) {
    if (!unpackDirs.includes(dirPath)) {
      unpackDirs.push(dirPath)
    }
    return true
  } else {
    return unpackDirs.some(unpackDir => dirPath.startsWith(unpackDir))
  }
}

module.exports.createPackage = async function (src, dest) {
  return module.exports.createPackageWithOptions(src, dest, {})
}

module.exports.createPackageWithOptions = async function (src, dest, options) {
  const globOptions = options.globOptions ? options.globOptions : {}
  globOptions.dot = options.dot === undefined ? true : options.dot

  const pattern = src + (options.pattern ? options.pattern : '/**/*')

  const [filenames, metadata] = await crawlFilesystem(pattern, globOptions)
  return module.exports.createPackageFromFiles(src, dest, filenames, metadata, options)
}

/**
 * Create an ASAR archive from a list of filenames.
 *
 * @param {string} src: Base path. All files are relative to this.
 * @param {string} dest: Archive filename (& path).
 * @param {array} filenames: List of filenames relative to src.
 * @param {object} metadata: Object with filenames as keys and {type='directory|file|link', stat: fs.stat} as values. (Optional)
 * @param {object} options: Options passed to `createPackageWithOptions`.
*/
module.exports.createPackageFromFiles = async function (src, dest, filenames, metadata, options) {
  if (typeof metadata === 'undefined' || metadata === null) { metadata = {} }
  if (typeof options === 'undefined' || options === null) { options = {} }

  src = path.normalize(src)
  dest = path.normalize(dest)
  filenames = filenames.map(function (filename) { return path.normalize(filename) })

  const filesystem = new Filesystem(src)
  const files = []
  const unpackDirs = []

  let filenamesSorted = []
  if (options.ordering) {
    const orderingFiles = (await fs.readFile(options.ordering)).toString().split('\n').map(line => {
      if (line.includes(':')) { line = line.split(':').pop() }
      line = line.trim()
      if (line.startsWith('/')) { line = line.slice(1) }
      return line
    })

    const ordering = []
    for (const file of orderingFiles) {
      const pathComponents = file.split(path.sep)
      let str = src
      for (const pathComponent of pathComponents) {
        str = path.join(str, pathComponent)
        ordering.push(str)
      }
    }

    let missing = 0
    const total = filenames.length

    for (const file of ordering) {
      if (!filenamesSorted.includes(file) && filenames.includes(file)) {
        filenamesSorted.push(file)
      }
    }

    for (const file of filenames) {
      if (!filenamesSorted.includes(file)) {
        filenamesSorted.push(file)
        missing += 1
      }
    }

    console.log(`Ordering file has ${((total - missing) / total) * 100}% coverage.`)
  } else {
    filenamesSorted = filenames
  }

  const handleFile = async function (filename) {
    if (!metadata[filename]) {
      metadata[filename] = await crawlFilesystem.determineFileType(filename)
    }
    const file = metadata[filename]

    let shouldUnpack
    switch (file.type) {
      case 'directory':
        if (options.unpackDir) {
          shouldUnpack = isUnpackedDir(path.relative(src, filename), options.unpackDir, unpackDirs)
        } else {
          shouldUnpack = false
        }
        filesystem.insertDirectory(filename, shouldUnpack)
        break
      case 'file':
        shouldUnpack = false
        if (options.unpack) {
          shouldUnpack = minimatch(filename, options.unpack, { matchBase: true })
        }
        if (!shouldUnpack && options.unpackDir) {
          const dirName = path.relative(src, path.dirname(filename))
          shouldUnpack = isUnpackedDir(dirName, options.unpackDir, unpackDirs)
        }
        files.push({ filename: filename, unpack: shouldUnpack })
        return filesystem.insertFile(filename, shouldUnpack, file, options)
      case 'link':
        filesystem.insertLink(filename)
        break
    }
    return Promise.resolve()
  }

  const insertsDone = async function () {
    await fs.mkdirp(path.dirname(dest))
    return disk.writeFilesystem(dest, filesystem, files, metadata)
  }

  const names = filenamesSorted.slice()

  const next = async function (name) {
    if (!name) { return insertsDone() }

    await handleFile(name)
    return next(names.shift())
  }

  return next(names.shift())
}

module.exports.statFile = function (archive, filename, followLinks) {
  const filesystem = disk.readFilesystemSync(archive)
  return filesystem.getFile(filename, followLinks)
}

module.exports.getRawHeader = function (archive) {
  return disk.readArchiveHeaderSync(archive)
}

module.exports.listPackage = function (archive, options) {
  return disk.readFilesystemSync(archive).listFiles(options)
}

module.exports.extractFile = function (archive, filename) {
  const filesystem = disk.readFilesystemSync(archive)
  return disk.readFileSync(filesystem, filename, filesystem.getFile(filename))
}

module.exports.extractAll = function (archive, dest) {
  const filesystem = disk.readFilesystemSync(archive)
  const filenames = filesystem.listFiles()

  // under windows just extract links as regular files
  const followLinks = process.platform === 'win32'

  // create destination directory
  fs.mkdirpSync(dest)

  for (const fullPath of filenames) {
    // Remove leading slash
    const filename = fullPath.substr(1)
    const destFilename = path.join(dest, filename)
    const file = filesystem.getFile(filename, followLinks)
    if (file.files) {
      // it's a directory, create it and continue with the next entry
      fs.mkdirpSync(destFilename)
    } else if (file.link) {
      // it's a symlink, create a symlink
      const linkSrcPath = path.dirname(path.join(dest, file.link))
      const linkDestPath = path.dirname(destFilename)
      const relativePath = path.relative(linkDestPath, linkSrcPath)
      // try to delete output file, because we can't overwrite a link
      try {
        fs.unlinkSync(destFilename)
      } catch {}
      const linkTo = path.join(relativePath, path.basename(file.link))
      fs.symlinkSync(linkTo, destFilename)
    } else {
      // it's a file, extract it
      const content = disk.readFileSync(filesystem, filename, file)
      fs.writeFileSync(destFilename, content)
      if (file.executable) {
        fs.chmodSync(destFilename, '755')
      }
    }
  }
}

module.exports.uncache = function (archive) {
  return disk.uncacheFilesystem(archive)
}

module.exports.uncacheAll = function () {
  disk.uncacheAll()
}
