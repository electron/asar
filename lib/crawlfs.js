'use strict'

const { promisify } = require('util')

const fs = require('./wrapped-fs')
const glob = promisify(require('glob'))

async function determineFileType (filename) {
  const stat = await fs.lstat(filename)
  if (stat.isFile()) {
    return { type: 'file', stat }
  } else if (stat.isDirectory()) {
    return { type: 'directory', stat }
  } else if (stat.isSymbolicLink()) {
    return { type: 'link', stat }
  }
}

module.exports = async function (dir, options) {
  const metadata = {}
  const crawled = await glob(dir, options)
  const results = await Promise.all(crawled.map(async filename => [filename, await determineFileType(filename)]))
  const links = []
  const filenames = results.map(([filename, type]) => {
    if (type) {
      metadata[filename] = type
      if (type.type === 'link') links.push(filename)
    }
    return filename
  }).filter((filename) => {
    // Newer glob can return files inside symlinked directories, to avoid
    // those appearing in archives we need to manually exclude theme here
    const exactLinkIndex = links.findIndex(link => filename === link)
    return links.every((link, index) => {
      if (index === exactLinkIndex) return true
      return !filename.startsWith(link)
    })
  })
  return [filenames, metadata]
}
module.exports.determineFileType = determineFileType
