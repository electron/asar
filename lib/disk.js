'use strict'

const fs = require('./wrapped-fs')
const path = require('path')
const pickle = require('chromium-pickle-js')

const Filesystem = require('./filesystem')
let filesystemCache = {}

async function copyFile (dest, src, filename) {
  const srcFile = path.join(src, filename)
  const targetFile = path.join(dest, filename)

  const [content, stats] = await Promise.all([fs.readFile(srcFile), fs.stat(srcFile), fs.mkdirp(path.dirname(targetFile))])
  return fs.writeFile(targetFile, content, { mode: stats.mode })
}

async function streamTransformedFile (originalFilename, outStream, transformed) {
  return new Promise((resolve, reject) => {
    const stream = fs.createReadStream(transformed ? transformed.path : originalFilename)
    stream.pipe(outStream, { end: false })
    stream.on('error', reject)
    stream.on('end', () => resolve())
  })
}

const writeFileListToStream = async function (dest, filesystem, out, list, metadata) {
  for (const file of list) {
    if (file.unpack) { // the file should not be packed into archive
      const filename = path.relative(filesystem.src, file.filename)
      await copyFile(`${dest}.unpacked`, filesystem.src, filename)
    } else {
      await streamTransformedFile(file.filename, out, metadata[file.filename].transformed)
    }
  }
  return out.end()
}

module.exports.writeFilesystem = async function (dest, filesystem, files, metadata) {
  const headerPickle = pickle.createEmpty()
  headerPickle.writeString(JSON.stringify(filesystem.header))
  const headerBuf = headerPickle.toBuffer()

  const sizePickle = pickle.createEmpty()
  sizePickle.writeUInt32(headerBuf.length)
  const sizeBuf = sizePickle.toBuffer()

  const out = fs.createWriteStream(dest)
  await new Promise((resolve, reject) => {
    out.on('error', reject)
    out.write(sizeBuf)
    return out.write(headerBuf, () => resolve())
  })
  return writeFileListToStream(dest, filesystem, out, files, metadata)
}

module.exports.readArchiveHeaderSync = function (archive) {
  const fd = fs.openSync(archive, 'r')
  let size
  let headerBuf
  try {
    const sizeBuf = Buffer.alloc(8)
    if (fs.readSync(fd, sizeBuf, 0, 8, null) !== 8) {
      throw new Error('Unable to read header size')
    }

    const sizePickle = pickle.createFromBuffer(sizeBuf)
    size = sizePickle.createIterator().readUInt32()
    headerBuf = Buffer.alloc(size)
    if (fs.readSync(fd, headerBuf, 0, size, null) !== size) {
      throw new Error('Unable to read header')
    }
  } finally {
    fs.closeSync(fd)
  }

  const headerPickle = pickle.createFromBuffer(headerBuf)
  const header = headerPickle.createIterator().readString()
  return { headerString: header, header: JSON.parse(header), headerSize: size }
}

module.exports.readFilesystemSync = function (archive) {
  if (!filesystemCache[archive]) {
    const header = this.readArchiveHeaderSync(archive)
    const filesystem = new Filesystem(archive)
    filesystem.header = header.header
    filesystem.headerSize = header.headerSize
    filesystemCache[archive] = filesystem
  }
  return filesystemCache[archive]
}

module.exports.uncacheFilesystem = function (archive) {
  if (filesystemCache[archive]) {
    filesystemCache[archive] = undefined
    return true
  }
  return false
}

module.exports.uncacheAll = function () {
  filesystemCache = {}
}

module.exports.readFileSync = function (filesystem, filename, info) {
  let buffer = Buffer.alloc(info.size)
  if (info.size <= 0) { return buffer }
  if (info.unpacked) {
    // it's an unpacked file, copy it.
    buffer = fs.readFileSync(path.join(`${filesystem.src}.unpacked`, filename))
  } else {
    // Node throws an exception when reading 0 bytes into a 0-size buffer,
    // so we short-circuit the read in this case.
    const fd = fs.openSync(filesystem.src, 'r')
    try {
      const offset = 8 + filesystem.headerSize + parseInt(info.offset)
      fs.readSync(fd, buffer, 0, info.size, offset)
    } finally {
      fs.closeSync(fd)
    }
  }
  return buffer
}
