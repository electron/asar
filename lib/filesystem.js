'use strict'

const fs = require('./wrapped-fs')
const os = require('os')
const path = require('path')
const { promisify } = require('util')
const stream = require('stream')
const getFileIntegrity = require('./integrity')

const UINT32_MAX = 2 ** 32 - 1

const pipeline = promisify(stream.pipeline)

class Filesystem {
  constructor (src) {
    this.src = path.resolve(src)
    this.header = { files: {} }
    this.offset = BigInt(0)
  }

  searchNodeFromDirectory (p) {
    let json = this.header
    const dirs = p.split(path.sep)
    for (const dir of dirs) {
      if (dir !== '.') {
        json = json.files[dir]
      }
    }
    return json
  }

  searchNodeFromPath (p) {
    p = path.relative(this.src, p)
    if (!p) { return this.header }
    const name = path.basename(p)
    const node = this.searchNodeFromDirectory(path.dirname(p))
    if (node.files == null) {
      node.files = {}
    }
    if (node.files[name] == null) {
      node.files[name] = {}
    }
    return node.files[name]
  }

  insertDirectory (p, shouldUnpack) {
    const node = this.searchNodeFromPath(p)
    if (shouldUnpack) {
      node.unpacked = shouldUnpack
    }
    node.files = {}
    return node.files
  }

  async insertFile (p, shouldUnpack, file, options) {
    const dirNode = this.searchNodeFromPath(path.dirname(p))
    const node = this.searchNodeFromPath(p)
    if (shouldUnpack || dirNode.unpacked) {
      node.size = file.stat.size
      node.unpacked = true
      node.integrity = await getFileIntegrity(p)
      return Promise.resolve()
    }

    let size

    const transformed = options.transform && options.transform(p)
    if (transformed) {
      const tmpdir = await fs.mkdtemp(path.join(os.tmpdir(), 'asar-'))
      const tmpfile = path.join(tmpdir, path.basename(p))
      const out = fs.createWriteStream(tmpfile)
      const readStream = fs.createReadStream(p)

      await pipeline(readStream, transformed, out)
      file.transformed = {
        path: tmpfile,
        stat: await fs.lstat(tmpfile)
      }
      size = file.transformed.stat.size
    } else {
      size = file.stat.size
    }

    // JavaScript cannot precisely present integers >= UINT32_MAX.
    if (size > UINT32_MAX) {
      throw new Error(`${p}: file size can not be larger than 4.2GB`)
    }

    node.size = size
    node.offset = this.offset.toString()
    node.integrity = await getFileIntegrity(p)
    if (process.platform !== 'win32' && (file.stat.mode & 0o100)) {
      node.executable = true
    }
    this.offset += BigInt(size)
  }

  insertLink (p) {
    const link = path.relative(fs.realpathSync(this.src), fs.realpathSync(p))
    if (link.substr(0, 2) === '..') {
      throw new Error(`${p}: file links out of the package`)
    }
    const node = this.searchNodeFromPath(p)
    node.link = link
    return link
  }

  listFiles (options) {
    const files = []

    const fillFilesFromMetadata = function (basePath, metadata) {
      if (!metadata.files) {
        return
      }

      for (const [childPath, childMetadata] of Object.entries(metadata.files)) {
        const fullPath = path.join(basePath, childPath)
        const packState = childMetadata.unpacked ? 'unpack' : 'pack  '
        files.push((options && options.isPack) ? `${packState} : ${fullPath}` : fullPath)
        fillFilesFromMetadata(fullPath, childMetadata)
      }
    }

    fillFilesFromMetadata('/', this.header)
    return files
  }

  getNode (p) {
    const node = this.searchNodeFromDirectory(path.dirname(p))
    const name = path.basename(p)
    if (name) {
      return node.files[name]
    } else {
      return node
    }
  }

  getFile (p, followLinks) {
    followLinks = typeof followLinks === 'undefined' ? true : followLinks
    const info = this.getNode(p)

    // if followLinks is false we don't resolve symlinks
    if (info.link && followLinks) {
      return this.getFile(info.link)
    } else {
      return info
    }
  }
}

module.exports = Filesystem
