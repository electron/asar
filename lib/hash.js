const crypto = require('crypto')
const fs = require('fs')
const stream = require('stream')
const { promisify } = require('util')

const ALGORITHMS = ['SHA256']

const pipeline = promisify(stream.pipeline)

async function hashFile (path) {
  const hashes = {}

  await Promise.all(ALGORITHMS.map(async (algo) => {
    const read = fs.createReadStream(path)
    const hash = crypto.createHash(algo)
    hash.setEncoding('base64')
    await pipeline(
      read,
      hash
    )

    hashes[algo] = hash.read()
  }))

  return hashes
}

module.exports = hashFile
