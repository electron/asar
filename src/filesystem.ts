import * as os from 'os';
import * as path from 'path';
import { promisify } from 'util';
import * as stream from 'stream';

import { FileIntegrity, getFileIntegrity } from './integrity';
import fs from './wrapped-fs';
import { CrawledFileType } from './crawlfs';

const UINT32_MAX = 2 ** 32 - 1;

const pipeline = promisify(stream.pipeline);

export type EntryMetadata = {
  unpacked?: boolean;
};

export type FilesystemDirectoryEntry = {
  files: Record<string, FilesystemEntry>;
} & EntryMetadata;

export type FilesystemFileEntry = {
  unpacked: boolean;
  executable: boolean;
  offset: string;
  size: number;
  integrity: FileIntegrity;
} & EntryMetadata;

export type FilesystemLinkEntry = {
  link: string;
} & EntryMetadata;

export type FilesystemEntry = FilesystemDirectoryEntry | FilesystemFileEntry | FilesystemLinkEntry;

export class Filesystem {
  private src: string;
  private header: FilesystemEntry;
  private headerSize: number;
  private offset: bigint;

  constructor(src: string) {
    this.src = path.resolve(src);
    this.header = { files: Object.create(null) };
    this.headerSize = 0;
    this.offset = BigInt(0);
  }

  getRootPath() {
    return this.src;
  }

  getHeader() {
    return this.header;
  }

  getHeaderSize() {
    return this.headerSize;
  }

  setHeader(header: FilesystemEntry, headerSize: number) {
    this.header = header;
    this.headerSize = headerSize;
  }

  searchNodeFromDirectory(p: string) {
    let json = this.header;
    const dirs = p.split(path.sep);
    for (const dir of dirs) {
      if (dir !== '.') {
        if ('files' in json) {
          if (!json.files[dir]) {
            json.files[dir] = { files: Object.create(null) };
          }
          json = json.files[dir];
        } else {
          throw new Error('Unexpected directory state while traversing: ' + p);
        }
      }
    }
    return json;
  }

  searchNodeFromPath(p: string) {
    p = path.relative(this.src, p);
    if (!p) {
      return this.header;
    }
    const name = path.basename(p);
    const node = this.searchNodeFromDirectory(path.dirname(p)) as FilesystemDirectoryEntry;
    if (!node.files) {
      node.files = Object.create(null);
    }
    if (!node.files[name]) {
      node.files[name] = Object.create(null);
    }
    return node.files[name];
  }

  insertDirectory(p: string, shouldUnpack: boolean) {
    const node = this.searchNodeFromPath(p) as FilesystemDirectoryEntry;
    if (shouldUnpack) {
      node.unpacked = shouldUnpack;
    }
    node.files = node.files || Object.create(null);
    return node.files;
  }

  async insertFile(
    p: string,
    shouldUnpack: boolean,
    file: CrawledFileType,
    options: {
      transform?: (filePath: string) => NodeJS.ReadWriteStream | void;
    } = {},
  ) {
    const dirNode = this.searchNodeFromPath(path.dirname(p)) as FilesystemDirectoryEntry;
    const node = this.searchNodeFromPath(p) as FilesystemFileEntry;
    if (shouldUnpack || dirNode.unpacked) {
      node.size = file.stat.size;
      node.unpacked = true;
      node.integrity = await getFileIntegrity(p);
      return Promise.resolve();
    }

    let size: number;

    const transformed = options.transform && options.transform(p);
    if (transformed) {
      const tmpdir = await fs.mkdtemp(path.join(os.tmpdir(), 'asar-'));
      const tmpfile = path.join(tmpdir, path.basename(p));
      const out = fs.createWriteStream(tmpfile);
      const readStream = fs.createReadStream(p);

      await pipeline(readStream, transformed, out);
      file.transformed = {
        path: tmpfile,
        stat: await fs.lstat(tmpfile),
      };
      size = file.transformed.stat.size;
    } else {
      size = file.stat.size;
    }

    // JavaScript cannot precisely present integers >= UINT32_MAX.
    if (size > UINT32_MAX) {
      throw new Error(`${p}: file size can not be larger than 4.2GB`);
    }

    node.size = size;
    node.offset = this.offset.toString();
    node.integrity = await getFileIntegrity(p);
    if (process.platform !== 'win32' && file.stat.mode & 0o100) {
      node.executable = true;
    }
    this.offset += BigInt(size);
  }

  insertLink(p: string) {
    const symlink = fs.readlinkSync(p);
    // /var => /private/var
    const parentPath = fs.realpathSync(path.dirname(p));
    const link = path.relative(fs.realpathSync(this.src), path.join(parentPath, symlink));
    if (link.startsWith('..')) {
      throw new Error(`${p}: file "${link}" links out of the package`);
    }
    const node = this.searchNodeFromPath(p) as FilesystemLinkEntry;
    node.link = link;
    return link;
  }

  listFiles(options?: { isPack: boolean }) {
    const files: string[] = [];

    const fillFilesFromMetadata = function (basePath: string, metadata: FilesystemEntry) {
      if (!('files' in metadata)) {
        return;
      }

      for (const [childPath, childMetadata] of Object.entries(metadata.files)) {
        const fullPath = path.join(basePath, childPath);
        const packState =
          'unpacked' in childMetadata && childMetadata.unpacked ? 'unpack' : 'pack  ';
        files.push(options && options.isPack ? `${packState} : ${fullPath}` : fullPath);
        fillFilesFromMetadata(fullPath, childMetadata);
      }
    };

    fillFilesFromMetadata('/', this.header);
    return files;
  }

  getNode(p: string, followLinks: boolean = true): FilesystemEntry {
    const node = this.searchNodeFromDirectory(path.dirname(p));
    const name = path.basename(p);
    if ('link' in node && followLinks) {
      return this.getNode(path.join(node.link, name));
    }
    if (name) {
      return (node as FilesystemDirectoryEntry).files[name];
    } else {
      return node;
    }
  }

  getFile(p: string, followLinks: boolean = true): FilesystemEntry {
    const info = this.getNode(p, followLinks);

    if (!info) {
      throw new Error(`"${p}" was not found in this archive`);
    }

    // if followLinks is false we don't resolve symlinks
    if ('link' in info && followLinks) {
      return this.getFile(info.link, followLinks);
    } else {
      return info;
    }
  }
}
