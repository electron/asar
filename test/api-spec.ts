import assert from 'assert';
import os from 'os';
import { FilesystemLinkEntry } from '../lib/filesystem';
import fs from '../lib/wrapped-fs';
import compDirs from './util/compareDirectories';
import compFileLists from './util/compareFileLists';
import { compFiles, isSymbolicLinkSync } from './util/compareFiles';
import transform from './util/transformStream';
import { verifySmartUnpack } from './util/verifySmartUnpack';

const asar = require('../src/asar');

async function assertPackageListEquals(actualList: string[], expectedFilename: string) {
  const expected = await fs.readFile(expectedFilename, 'utf8');
  return compFileLists(actualList.join('\n'), expected);
}

describe('api', function () {
  it('should create archive from directory', async () => {
    await asar.createPackage('test/input/packthis/', 'tmp/packthis-api.asar');
    return compFiles('tmp/packthis-api.asar', 'test/expected/packthis.asar');
  });
  it.ifWindows('should create archive with windows-style path separators', async () => {
    await asar.createPackage('test\\input\\packthis\\', 'tmp\\packthis-api.asar');
    return compFiles('tmp/packthis-api.asar', 'test/expected/packthis.asar');
  });
  it('should create archive from directory (without hidden files)', async () => {
    await asar.createPackageWithOptions(
      'test/input/packthis/',
      'tmp/packthis-without-hidden-api.asar',
      { dot: false },
    );
    return compFiles(
      'tmp/packthis-without-hidden-api.asar',
      'test/expected/packthis-without-hidden.asar',
    );
  });
  it('should create archive from directory (with transformed files)', async () => {
    await asar.createPackageWithOptions(
      'test/input/packthis/',
      'tmp/packthis-api-transformed.asar',
      { transform },
    );
    return compFiles(
      'tmp/packthis-api-transformed.asar',
      'test/expected/packthis-transformed.asar',
    );
  });
  it('should create archive from directory (with nothing packed)', async () => {
    const out = 'tmp/packthis-api-unpacked.asar';
    await asar.createPackageWithOptions('test/input/packthis/', out, {
      unpackDir: '**',
    });
    await verifySmartUnpack(out);
  });
  it('should list files/dirs in archive', async () => {
    return assertPackageListEquals(
      asar.listPackage('test/input/extractthis.asar', { isPack: false }),
      'test/expected/extractthis-filelist.txt',
    );
  });
  it('should list files/dirs in archive with option', async () => {
    return assertPackageListEquals(
      asar.listPackage('test/input/extractthis-unpack-dir.asar', { isPack: true }),
      'test/expected/extractthis-filelist-with-option.txt',
    );
  });
  it('should extract a text file from archive', async () => {
    const actual = asar
      .extractFile('test/input/extractthis.asar', 'dir1/file1.txt')
      .toString('utf8');
    const expected = await fs.readFile('test/expected/extractthis/dir1/file1.txt', 'utf8');
    return compFileLists(actual, expected);
  });
  it('should extract a binary file from archive', async () => {
    const actual = asar.extractFile('test/input/extractthis.asar', 'dir2/file2.png');
    const expected = await fs.readFile('test/expected/extractthis/dir2/file2.png');
    return assert.strictEqual(actual.toString(), expected.toString());
  });
  it('should extract a binary file from archive with unpacked files', async () => {
    const actual = asar.extractFile('test/input/extractthis-unpack.asar', 'dir2/file2.png');
    const expected = await fs.readFile('test/expected/extractthis/dir2/file2.png');
    return assert.strictEqual(actual.toString(), expected.toString());
  });
  it('should extract an archive', async () => {
    asar.extractAll('test/input/extractthis.asar', 'tmp/extractthis-api/');
    return compDirs('tmp/extractthis-api/', 'test/expected/extractthis');
  });
  it('should extract an archive with unpacked files', async () => {
    asar.extractAll('test/input/extractthis-unpack.asar', 'tmp/extractthis-unpack-api/');
    return compDirs('tmp/extractthis-unpack-api/', 'test/expected/extractthis');
  });
  it('should extract a binary file from archive with unpacked files', async () => {
    const actual = asar.extractFile('test/input/extractthis-unpack-dir.asar', 'dir1/file1.txt');
    const expected = await fs.readFile('test/expected/extractthis/dir1/file1.txt');
    assert.strictEqual(actual.toString(), expected.toString());
  });
  it('should extract an archive with unpacked dirs', async () => {
    asar.extractAll('test/input/extractthis-unpack-dir.asar', 'tmp/extractthis-unpack-dir-api/');
    return compDirs('tmp/extractthis-unpack-dir-api/', 'test/expected/extractthis');
  });

  // We don't extract symlinks on Windows, so skip these tests
  it.ifNotWindows('should extract an archive with symlink', async () => {
    assert.strictEqual(isSymbolicLinkSync('test/input/packthis-with-symlink/real.txt'), true);
    await asar.createPackageWithOptions(
      'test/input/packthis-with-symlink/',
      'tmp/packthis-with-symlink.asar',
      { dot: false },
    );
    asar.extractAll('tmp/packthis-with-symlink.asar', 'tmp/packthis-with-symlink/');
    return compFiles(
      'tmp/packthis-with-symlink/real.txt',
      'test/input/packthis-with-symlink/real.txt',
    );
  });
  it.ifNotWindows('should extract an archive with symlink having the same prefix', async () => {
    assert.strictEqual(
      isSymbolicLinkSync('test/input/packthis-with-symlink-same-prefix/real.txt'),
      true,
    );
    await asar.createPackageWithOptions(
      'test/input/packthis-with-symlink-same-prefix/',
      'tmp/packthis-with-symlink-same-prefix.asar',
      { dot: false },
    );
    asar.extractAll(
      'tmp/packthis-with-symlink-same-prefix.asar',
      'tmp/packthis-with-symlink-same-prefix/',
    );
    return compFiles(
      'tmp/packthis-with-symlink-same-prefix/real.txt',
      'test/input/packthis-with-symlink-same-prefix/real.txt',
    );
  });
  it.ifNotWindows('should not extract an archive with a bad symlink', async () => {
    assert.throws(() => {
      asar.extractAll('test/input/bad-symlink.asar', 'tmp/bad-symlink/');
    });
  });
  it('should handle multibyte characters in paths', async () => {
    await asar.createPackageWithOptions(
      'test/input/packthis-unicode-path/',
      'tmp/packthis-unicode-path.asar',
      {
        globOptions: {
          nosort: true,
        },
      },
    );
    return compFiles('tmp/packthis-unicode-path.asar', 'test/expected/packthis-unicode-path.asar');
  });
  it('should extract a text file from archive with multibyte characters in path', async () => {
    const actual = asar
      .extractFile('test/expected/packthis-unicode-path.asar', 'dir1/女の子.txt')
      .toString('utf8');
    const expected = await fs.readFile('test/input/packthis-unicode-path/dir1/女の子.txt', 'utf8');
    return compFileLists(actual, expected);
  });
  it('should create files/directories whose names are properties of Object.prototype', async () => {
    await asar.createPackage(
      'test/input/packthis-object-prototype/',
      'tmp/packthis-object-prototype.asar',
    );
    return compFiles(
      'tmp/packthis-object-prototype.asar',
      'test/expected/packthis-object-prototype.asar',
    );
  });
  it('should extract files/directories whose names are properties of Object.prototype', () => {
    asar.extractAll(
      'test/expected/packthis-object-prototype.asar',
      'tmp/packthis-object-prototype/',
    );
    return compDirs('test/input/packthis-object-prototype/', 'tmp/packthis-object-prototype');
  });
  it('should export all functions also in the default export', () => {
    const topLevelFunctions = Object.keys(asar).filter((key) => typeof asar[key] === 'function');
    const defaultExportFunctions = Object.keys(asar.default).filter(
      (key) => typeof asar.default[key] === 'function',
    );

    assert.deepStrictEqual(topLevelFunctions, defaultExportFunctions);
  });
  it('should stat a symlinked file', async () => {
    const stats = asar.statFile(
      'test/input/stat-symlink.asar',
      'real.txt',
      true,
    ) as FilesystemLinkEntry;
    return assert.strictEqual(stats.link, undefined);
  });
});
