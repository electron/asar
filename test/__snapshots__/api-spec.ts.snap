// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`api should create archive from directory (with nothing packed) 1`] = `
{
  "files": {
    ".hiddenfile.txt": {
      "integrity": {
        "algorithm": "SHA256",
        "blockSize": 4194304,
        "blocks": [
          "d400d9a74f67e724a4793173d73e1403bd6b734930632f4eee50a3a5a1e5a478",
        ],
        "hash": "d400d9a74f67e724a4793173d73e1403bd6b734930632f4eee50a3a5a1e5a478",
      },
      "size": 19,
      "unpacked": true,
    },
    "dir1": {
      "files": {
        "file1.txt": {
          "integrity": {
            "algorithm": "SHA256",
            "blockSize": 4194304,
            "blocks": [
              "420149d3f852894ba7f32e9d3ec7d8919ee2451724bf2580b0186cd373bd6d82",
            ],
            "hash": "420149d3f852894ba7f32e9d3ec7d8919ee2451724bf2580b0186cd373bd6d82",
          },
          "size": 9,
          "unpacked": true,
        },
      },
      "unpacked": true,
    },
    "dir2": {
      "files": {
        "file2.png": {
          "integrity": {
            "algorithm": "SHA256",
            "blockSize": 4194304,
            "blocks": [
              "cc402b796dc92b2b1f3a6d09515003d8400e63d8acaffc967e49c0cf015fcffe",
            ],
            "hash": "cc402b796dc92b2b1f3a6d09515003d8400e63d8acaffc967e49c0cf015fcffe",
          },
          "size": 182,
          "unpacked": true,
        },
        "file3.txt": {
          "integrity": {
            "algorithm": "SHA256",
            "blockSize": 4194304,
            "blocks": [
              "a665a45920422f9d417e4867efdc4fb8a04a1f3fff1fa07e998e86f7f7a27ae3",
            ],
            "hash": "a665a45920422f9d417e4867efdc4fb8a04a1f3fff1fa07e998e86f7f7a27ae3",
          },
          "size": 3,
          "unpacked": true,
        },
      },
      "unpacked": true,
    },
    "emptyfile.txt": {
      "integrity": {
        "algorithm": "SHA256",
        "blockSize": 4194304,
        "blocks": [
          "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855",
        ],
        "hash": "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855",
      },
      "size": 0,
      "unpacked": true,
    },
    "file0.txt": {
      "integrity": {
        "algorithm": "SHA256",
        "blockSize": 4194304,
        "blocks": [
          "41a978be88ff87a57308bf234106a57ebf9aa0971e2101e75113547a815be72b",
        ],
        "hash": "41a978be88ff87a57308bf234106a57ebf9aa0971e2101e75113547a815be72b",
      },
      "size": 13,
      "unpacked": true,
    },
  },
}
`;

exports[`api should create archive from directory (with nothing packed) 2`] = `
[
  {
    "content": "file one.",
    "name": "dir1/file1.txt",
  },
  "dir2/file2.png",
  {
    "content": "123",
    "name": "dir2/file3.txt",
  },
  "dir1",
  "dir2",
  {
    "content": "This file is hidden",
    "name": ".hiddenfile.txt",
  },
  {
    "content": "",
    "name": "emptyfile.txt",
  },
  {
    "content": "file0 content",
    "name": "file0.txt",
  },
]
`;
